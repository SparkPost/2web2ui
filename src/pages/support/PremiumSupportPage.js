import React, { Component } from 'react';
import { connect } from 'react-redux';
import { reduxForm } from 'redux-form';
import { createTicket } from 'src/actions/support';
import { showAlert } from 'src/actions/globalAlert';
import { PageLink } from 'src/components/links';
import { Button, Page, Panel } from 'src/components/matchbox';
import { Form } from 'src/components/form';
import PremiumSupportFields from './components/PremiumSupportFields';
import { generateMessage } from './helpers/formHelpers';

export class PremiumSupportPage extends Component {
  handleTicketCreate = values =>
    this.props.createTicket({
      message: generateMessage(values),
      issueType: 'Vetting',
      subject: 'AWS Premium Vetting',
    });

  componentDidUpdate(prevProps) {
    const { submitSucceeded, showAlert, history, ticketId, createError } = this.props;

    if (!prevProps.submitSucceeded && submitSucceeded && !createError) {
      history.push('/account/billing');
      return showAlert({
        type: 'success',
        message: `Request Submitted - Ticket # ${ticketId}`,
        autoDismiss: false,
      });
    }
  }

  render() {
    const { handleSubmit, submitting } = this.props;

    return (
      <Page
        breadcrumbAction={{
          as: PageLink,
          to: '/account/billing',
          content: 'Back to billing',
        }}
      >
        <Panel.LEGACY title="Request Premium Support" sectioned accent>
          <Form onSubmit={handleSubmit(this.handleTicketCreate)}>
            <PremiumSupportFields />
            <Button submit color="orange" disabled={submitting}>
              {submitting ? 'Submitting...' : 'Submit Request'}
            </Button>
          </Form>
        </Panel.LEGACY>
      </Page>
    );
  }
}

const formName = 'AwsPremiumSupportForm';
const mapStateToProps = state => ({
  ticketId: state.support.ticketId,
  createError: state.support.createError,
});
const formOptions = { form: formName };
export default connect(mapStateToProps, { createTicket, showAlert })(
  reduxForm(formOptions)(PremiumSupportPage),
);
