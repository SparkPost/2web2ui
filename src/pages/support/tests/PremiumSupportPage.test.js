import React from 'react';
import { shallow } from 'enzyme';
import { PremiumSupportPage } from '../PremiumSupportPage';

jest.mock('../helpers/formHelpers');

const props = {
  loading: false,
  keys: ['an array of keys'],
  handleSubmit: jest.fn(a => a),
  createTicket: jest.fn(),
  submitting: false,
  history: { push: jest.fn() },
  showAlert: jest.fn(),
};

let wrapper;

describe('PremiumSupportPage', () => {
  beforeEach(() => {
    wrapper = shallow(<PremiumSupportPage {...props} />);
  });

  it('default state', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('renders submitting state correctly', () => {
    wrapper.setProps({ submitting: true });
    expect(wrapper.find('Button')).toMatchSnapshot();
  });

  it('submits Form correctly', () => {
    wrapper.find('Form').simulate('submit');
    expect(props.handleSubmit).toHaveBeenCalledTimes(1);
    expect(props.createTicket.mock.calls[0]).toMatchSnapshot();
  });

  it('handles submit success', () => {
    wrapper.setProps({ submitSucceeded: true, ticketId: '123' });
    expect(props.history.push).toHaveBeenCalledWith('/account/billing');
    expect(props.showAlert.mock.calls[0]).toMatchSnapshot();
  });

  it('handles submit error case', () => {
    wrapper.setProps({ submitSucceeded: true, createError: 'error message' });
    expect(props.history.push).not.toHaveBeenCalled();
    expect(props.showAlert).not.toHaveBeenCalled();
  });
});
