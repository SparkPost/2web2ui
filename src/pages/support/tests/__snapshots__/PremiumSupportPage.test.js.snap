// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PremiumSupportPage default state 1`] = `
<Page
  breadcrumbAction={
    Object {
      "as": [Function],
      "content": "Back to billing",
      "to": "/account/billing",
    }
  }
>
  <Panel.LEGACY
    accent={true}
    sectioned={true}
    title="Request Premium Support"
  >
    <Form
      id="premium-support-form"
      onSubmit={[Function]}
    >
      <PremiumSupportFields />
      <Button
        color="orange"
        disabled={false}
        submit={true}
      >
        Submit Request
      </Button>
    </Form>
  </Panel.LEGACY>
</Page>
`;

exports[`PremiumSupportPage handles submit success 1`] = `
Array [
  Object {
    "autoDismiss": false,
    "message": "Request Submitted - Ticket # 123",
    "type": "success",
  },
]
`;

exports[`PremiumSupportPage renders submitting state correctly 1`] = `
<Button
  color="orange"
  disabled={true}
  submit={true}
>
  Submitting...
</Button>
`;

exports[`PremiumSupportPage submits Form correctly 1`] = `
Array [
  Object {
    "issueType": "Vetting",
    "message": undefined,
    "subject": "AWS Premium Vetting",
  },
]
`;
